### PR TITLE
Module to filter mixed detector steps

### PR DESCRIPTION
**Pull request by edcallaghan**
This PR implements a tool-backed system for filtering detector step collections as appear in the standard mixing workflow, principally to avoid double-counting when simulating primaries which may be produced in upstream G4 simulation (e.g., IPA DIOs). There is currently an ugly wart present, the `InnerProtonAbsorberPseudoVolumeLookupTool`, which is needed in the immediate-term to work around missing `PhysicalVolumeInfo` data products, which will be retired after future datasets are produced. A sibling PR in `Production` is https://github.com/Mu2e/Production/pull/361.

New module:
- `MixingFilter`: Takes each a `CaloShowerStepCollection`, `CrvStepCollection`, and `StrawGasStepCollection` with a common `art::InputTag`, as appropriate for the standard mixing workflow, and filters these collections according to selections defined in a common "tool" (as in `art`). If no tool is configured, a "trivial filter" is applied to pass all steps. The selected steps in each subsystem are deep-copied into new collections and inserted into the event with no `art::Ptr` remapping; this is not necessary as the scope of the module is to filter only detector steps, which have no pointers to each other. I was somewhat surprised to find that the overhead of running the trivial filter is negligible, _well_ below 1% of the actual mixing module itself.

New tools:
- `DetectorStepSelectionTool`: Abstract class defining the interface for selections to be applied in the above, namely three `Select` functions which return a boolean specifying whether a step should be retained. Active selection was chosen over active rejection because the higher-level module must instantiate a new collection, which requires one full pass through the data; subsequent removal of elements would necessitate a second pass. Since the implementation is active-selection, it is more clear for the selection interface to be active-selection as well.
- `UniversalDetectorStepSelectionTool`: Tool implementing the trivial selection of passing all detector steps.
- `ProcessVolumeDetectorStepAntiSelectionTool`: Tool implementing a selection which omits steps which were taken by particles produced by a specific process in a specific volume, or a descendant thereof. Because of the latter detail, which often requires traversing the full tracking ancestry of a given particle, the overhead of this tool is non-negligible, landing in the low 10s-of-%, again with respect to the mixing module alone. That being said, the largest mixing stream by far comes from neutrals, which I don't know of a current, non-trivial, use of this system for. For the immediate case of interest, IPA DIOs, the overhead added to full-event processing is on the order of 0.1%.
  The querying of a particle's starting volume should be performed via a `PhysicalVolumeMultiHelper`-mediated query of a `PhysicalVolumeInfoMultiCollection` object; unfortunately, the latter data product was dropped from the relevant `MDC2020p` datasets, so this cannot, at present, be utilized. As a stop-gap to allow for proceeding with the IPA DIO workflow, this query is replaced with an effective query of whether a particle came from the current IPA geometry, which is described below. This is intended to be temporary, and is implemented in such a way that it's future removal will not require changing the configuration interface for this tool.

New, but temporary, functionality:
- `InnerProtonAbsorberPseudoVolumeLookupTool`: A tool which, using the `GeometryService` representation of the proton absorbers, determines whether a point is in the cylindrical shell defined by the IPA. If so, it returns the G4 volume name associated with the IPA (as a query of into a `PhysicalVolumeInfoMultiCollection` would), and if not, it returns a configurable alternative. Because this is a stop-gap, its configuration is hard-coded in the tool in which it is used, with an alternative volume name of poetic nonsense which will not match any actual G4 volume name in our application. When new simulations are produced with `PhysicalVolumeInfoMultiCollection`s retained, this tool will be retired and exchanged for queries into the latter.

### CODE REVIEW

**Review by brownd1978 — COMMENTED** [review]
I don't see where a momentum cut or creation process cut is made. Is that outside this PR?

File: EventMixing/inc/InnerProtonAbsorberPseudoVolumeLookupTool.hh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#include "fhiclcpp/types/Name.h"
+
+// mu2e
+#include "Offline/GeometryService/inc/DetectorSystem.hh"
+#include "Offline/GeometryService/inc/GeomHandle.hh"
+#include "Offline/MCDataProducts/inc/SimParticle.hh"
+#include "Offline/MECOStyleProtonAbsorberGeom/inc/MECOStyleProtonAbsorber.hh"
+
+namespace mu2e{
+  class InnerProtonAbsorberPseudoVolumeLookupTool{
```
[the remark this answers is not in the document]
Comment (edcallaghan) [reply EventMixing/inc/InnerProtonAbsorberPseudoVolumeLookupTool.hh:30]: ``Pseudo'' here refers to being a "volume lookup." I.e. this tool really determines the boolean of whether a particle came from the IPA or not, but with an interface that looks like it's doing a G4 volume name lookup.

**Comment by edcallaghan** [comment]
@brownd1978 
> I don't see where a momentum cut or creation process cut is made. Is that outside this PR?

`InnerProtonAbsorberPseudoVolumeLookupTool_tool.cc` lines lines 35 to 48 contain the core cut logic.

**Comment by edcallaghan** [comment]
@FNALbuild run build test with Mu2e/Production#361

**Comment by gaponenko** [comment]
Hi Ed,

Can you please comment on the design of this system?   I struggle to understand the point of having tools here.   Why not just use three dedicated modules to filter the 3 collections (CaloShowerSteps, CrvSteps, StrawGasSteps)?    The modules can still use templated code.   The tools do not seem to be re-used from multiple places, so they only add complexity here.

Andrei

**Comment by edcallaghan** [comment]
@gaponenko Here is the reasoning that led to the current implementation. The reason that I've introduced tools is in the third bullet.

  1) All of the mixed detector steps should be filtered according to the same rule (because all of the detectors should respond to the same physics). Whether the logic for that rule is written in a module or in a tool, it should be common to three collections.
  2) Given (1), having multiple modules would just triplicate the configuration and make maintenance more difficult. Without any filtering, the provenances of the three detector steps are the same (the mixing module), but partitioned by their origin stream (`EleBeam`, `MuBeam`, ...); using a single filtering module maintains that pattern. just with a substitution of `Mixing` -> `Filtering`.
  3) The logic of the filtering rule is encapsulated in a tool to i) obviate the need for redundant loops over the collections to be rewritten when a new use-case comes up in the future, and ii) simplify the production configuration and its maintenance, given that the configuration will change for different primaries. If we use "bare" modules, then when we switch from one module to another to simulate different primaries, that would entail a change a in provenance, which would need to be accommodated in all downstream configuration. We would react by trying to short-circuit that with some kind of aliasing of the modules in fcl, but then that adds more hair to the configuration and may potentially lead to confusing situations  where people will mistakenly associate a module label with a different module from what they were expecting.
    Deferring the core filtering logic to tools lets the "basic" configuration be a simple extension of how it currently looks (4 filter modules, each an analog to one of the 4 mixing modules; and they are each the same module, just as each of the mixing modules are the same), and allows to enable a nontrivial filter for a real use case by just overriding a single field in that module, to configure the tool, without any change in provenance or relabelling of modules.

  The one exception is the `InnerProtonAbsorberPseudoVolume...` thing, which is just a temporary hack and has no long-term structural consequences.

  Happy to hear more thoughts on these.

**Comment by edcallaghan** [comment]
@FNALbuild run build test

**Comment by gaponenko** [comment]
Ed, thank you for the explanation.   I have not appreciated that you have a single fcl configuration for the C++ code that does the filtering, not 3 pieces of configuration that need to be kept identical.   Then there is still a choice of whether that C++ code is in a tool or a module.  It sounds like the preference is to minimize future changes in the sequence of modules run in production, so you use a fixed module and keep the possibility of changing the filtering by modifying the tools.

**Comment by edcallaghan** [comment]
@gaponenko That is correct (and more succinct than how I put it!).

**Comment by edcallaghan** [comment]
@FNALbuild run build test with Mu2e/Production#361

**Review by brownd1978 — APPROVED** [review]
This looks complete for the IPA. We talked too about the stopping target filter, which should have a very similar implementation. Are you planning to include that in this PR or is that future work?

**Comment by brownd1978** [comment]
Could you please look into the clangtidy error? Those generally are real errors.

**Comment by edcallaghan** [comment]
@brownd1978 I will push an update to allow the same filter to work for the stopping target. Some of the clang-tidys are due to the new stuff here, which I will deal with; will take a stab at the others, but some may be out of scope.

**Comment by brownd1978** [comment]
The only clangtidy I care about is the error, the warnings are benign.

**Comment by edcallaghan** [comment]
@FNALbuild run build test with Mu2e/Production#361

**Comment by edcallaghan** [comment]
@brownd1978 looks good now. credit to @gaponenko for giving me the easy solution.

**Comment by edcallaghan** [comment]
Ah, but let's hold off until the stopping target volume lookup is added. Will be soon.

**Comment by edcallaghan** [comment]
@FNALbuild run build test with Mu2e/Production#361

**Comment by edcallaghan** [comment]
Ok, I went on a bit of a goose chase due to misunderstanding how the DIO pileup was generated, but this is now ready to merge. I have verified that we can identify properly identify DIOs in both the IPA and stopping target, and the treatment of volumes is future-proofed for when we have proper volume info and the latter will be distributed over the multiple foils.

**Comment by edcallaghan** [comment]
Snuck one more small one in, which exchanges a one-sided cut on momentum for a dual-optional two-sided cut on energy. This matches how the DIO and IPA primary generators are configured, so that the configurations can be matched exactly.

**Comment by edcallaghan** [comment]
@FNALbuild run build test with Mu2e/Production#361

**Pull request merged**